### ISSUE TITLE
Use GDK to retrieve desktop info and possibly window title

### ISSUE THREAD

**Issue by Lokaltog** [Lokaltog/candybar#48]
To avoid having XCB as a dependency (to support Wayland) for the desktop list and window title widgets, we can use GDK to get desktop info. [Here](http://www.futuredesktop.org/tmp/desk-test.c) is a source file that demonstrates using `gdk_property_get` to get WM properties like the desktop list, current desktop, etc.

**Comment by bbenne10** [comment]
I'd LOVE to see this implemented. I'm going to look over the code and see how well I might be able to implement it tonight.

**Comment by bbenne10** [comment]
annnnnnd I can tell I won't get this implemented tonight. I've got quite a bit of reading on in here. If anyone gets to this before me (I have a rather busy week), you're not stepping on my toes :)

**Comment by Lokaltog** [comment]
Allright! I'm not planning to do this right now, but I've stumbled across another issue that needs to be resolved before this will work without XCB. Currently XCB is required because we also subscribe to event notifications (window changed, desktop changed, desktop list update, etc.) and I'm not sure if GDK provides access to this low-level functionality. There's the [X-specific GDK extensions](https://developer.gnome.org/gdk3/stable/gdk3-X-Window-System-Interaction.html) and [gdk_window_add_filter](https://developer.gnome.org/gdk3/stable/gdk3-Windows.html#gdk-window-add-filter), if anyone is up to the task of digging through the docs to see if it's possible to do this in GDK it would be awesome.

**Comment by bbenne10** [comment]
The subscription model is what I was hung up on as well. `gdk_window_add_filter` sounds like a good approach, but we'll have to figure out a way to dispatch events we're not interested in (and we may need multiple filters if we want to do this in plugins, since each plugin is responsible for a single "event"). 

Wouldn't we want to avoid the X11 specific GDK extensions to remain Wayland agnostic?

**Comment by Lokaltog** [comment]
Ideally yes, but I'm not sure if it's even possible yet. I don't think the Wayland API provides a lot of the functionality required for this yet (subscribing to events, getting window data, etc), so what I think we should do for now is to provide this functionality through GDK to avoid XCB, and if/when Wayland supports these features in the future we can implement them for Wayland too. It's not a permanent solution, and I don't see how we can be completely wayland agnostic if we want to provide features that are currently only available in X.

**Comment by Lokaltog** [comment]
After reading and working a bit more on this issue I'm not sure if this will work, and it may be a lot easier to just make one widget for X and one widget for Wayland (e.g. `desktops_x`, `desktops_wl`, or make it a setting in `config.json`). The lower-level GDK functions appear to basically just be wrappers around Xlib functions currently, so we might as well just use XCB to subscribe to events and read window properties instead of trying to make the GDK functions work.

As an example, I initially tried using GDK to fetch the desktop names in 0f55fb2c34dae7cd00a8c3418e1e0677a14c5958, but the GDK function for reading UTF-8 lists from EWMH properties managed to screw it up completely (random values inside the strings, splitting names at wrong points, invalid desktop names, etc.) The xcb-wm function just worked right away so I ended up using that instead.

tl;dr When wayland is ready and if it ends up providing an API for reading window properties like X does, I think we should create wayland-specific widgets or functions for these features.
